### DISCUSSION TITLE
Assemble instructions with correct base IP

### DISCUSSION THREAD

**Discussion by qwerty01** [icedland/iced#288] [Q&A] [answered]
I've been trying to understand the crate's `CodeAssembler` and I'm having a little trouble on the final assembling step. I tried to duplicate the example disassemble code, and it seems to be ignoring the base IP I set in the assemble options:

```rust
    let mut asm = CodeAssembler::new(64)?;

    asm.mov(ptr(rsp + 0x10), rbx)?;
    asm.mov(ptr(rsp + 0x18), rsi)?;
    asm.push(rbp)?;
    asm.push(rdi)?;
    asm.push(r14)?;
    asm.lea(rbp, ptr(rsp-0x100i32))?;
    asm.sub(rsp, 0x200)?;
    asm.mov(rax, ptr(0x7ffac4524e0u64))?;
    asm.xor(rax, rsp)?;
    asm.mov(ptr(rbp + 0xf0), rax)?;
    asm.mov(r8, ptr(0x7ffac474f208u64))?; // This is the line the assembler is complaining about
    asm.lea(rax, ptr(0x7ffac46f4a58u64))?;
    asm.xor(edi, edi)?;

    let address = 0x7FFAC46ACDA4;
    let result = asm.assemble_options(address, BlockEncoderOptions::NONE)?;
```
but when this code is run, it returns the following error:
```
Operand 1: Displacement must fit in an i32 : 0x0 mov r8,[7FFAC474F208h]
```
the address of the `mov` instruction it's complaining about should be `0x7FFAC46ACDD2`, or somewhere close to that if it didn't assemble exactly the same as the example, so the displacement should be `0xa2436`, which is well within the limits of an i32. Is there some other option that needs to be set in order for it to assemble with the proper IP value for that instruction or some other issue that could be causing it to assemble with `0x0` as the IP?

**Comment by wtfsck** [comment, marked as answer]
Sounds like https://github.com/icedland/iced/issues/212 and if so, the workaround is https://github.com/icedland/iced/issues/209#issuecomment-950273687

**Reply by qwerty01** [reply]
Ah yep that's it, I was having a hard time finding what to search in the issues, thanks
